const-oid/src/lib.rs

```diff
@@ -21,50 +25,183 @@
+    ($bool:expr, $msg:expr) => {
+        [$msg][!$bool as usize]
+    };
 }
 
 impl ObjectIdentifier {
-    /// Create a new OID
-    #[allow(clippy::no_effect)]
-    pub const fn new(nodes: &'static [u32]) -> Self {
-        // TODO(tarcieri): replace this with const panic when OIDs are invalid
-        // See: https://github.com/rust-lang/rust/issues/51999
-        let mut is_invalid = nodes.len() <= 2;
-
-        match nodes[0] {
-            0..=2 => (),
-            _ => is_invalid = true,
+    /// Create an OID from a slice of integers.
+    ///
+    /// # Panics
+    ///
+    /// To enable `const fn` usage and work around current limitations thereof,
+    /// this method panics in the event the OID is malformed.
+    ///
+    /// For that reason this method is not recommended except for use in
+    /// constants (where it will generate a compiler error instead).
+    ///
+    /// In order for an OID to be valid, it must meet the following criteria:
+    ///
+    /// - The OID MUST have at least 3 nodes
+    /// - The OID MUST NOT have more nodes than the [`MAX_NODES`] constant
+    /// - The first node MUST be within the range 0-2
+    /// - The second node MUST be within the range 0-39
+    pub const fn new(nodes: &[u32]) -> Self {
+        const_assert!(nodes.len() >= 3, "OID too short (minimum 3 nodes)");
+        const_assert!(
+            nodes.len() <= MAX_NODES,
+            "OID too long (internal limit reached)"
+        );
+        const_assert!(
+            nodes[0] <= FIRST_NODE_MAX,
+            "invalid first node (must be 0-2)"
+        );
+        const_assert!(
+            nodes[1] <= SECOND_NODE_MAX,
+            "invalid second node (must be 0-39)"
+        );
+
+        // TODO(tarcieri): use `const_mut_ref` when stable.
+        // See: <https://github.com/rust-lang/rust/issues/57349>
+        let n = match nodes.len() {
+            3 => [nodes[0], nodes[1], nodes[2], 0, 0, 0, 0, 0, 0, 0],
+            4 => [nodes[0], nodes[1], nodes[2], nodes[3], 0, 0, 0, 0, 0, 0],
+            5 => [
+                nodes[0], nodes[1], nodes[2], nodes[3], nodes[4], 0, 0, 0, 0, 0,
+            ],
+            6 => [
+                nodes[0], nodes[1], nodes[2], nodes[3], nodes[4], nodes[5], 0, 0, 0, 0,
+            ],
+            7 => [
+                nodes[0], nodes[1], nodes[2], nodes[3], nodes[4], nodes[5], nodes[6], 0, 0, 0,
+            ],
+            8 => [
+                nodes[0], nodes[1], nodes[2], nodes[3], nodes[4], nodes[5], nodes[6], nodes[7], 0,
+                0,
+            ],
+            9 => [
+                nodes[0], nodes[1], nodes[2], nodes[3], nodes[4], nodes[5], nodes[6], nodes[7],
+                nodes[8], 0,
+            ],
+            10 => [
+                nodes[0], nodes[1], nodes[2], nodes[3], nodes[4], nodes[5], nodes[6], nodes[7],
+                nodes[8], nodes[9],
+            ],
+            _ => [0u32; MAX_NODES], // Checks above prevent this case, but makes the compiler happy
```

Comment:
Use `unreachable!()` here?

Reply:
It's a `const fn`, so panic is unsupported.

Best I could do is `unreachable_unchecked()`, but that seems like a needless use of `unsafe` in an otherwise `forbid(unsafe_code)` crate.